### PR TITLE
Allow getModuleConfig() to execute before module instantiation

### LINKED ISSUE #143: Allow module configuration object to be available ahead of module creation
Use Case:

I have two mutually exclusive behaviors that need to be added to the module depending on the context.

```
Box.Application.addModule('foo', function(context) {
    var behaviors = ['bar'];

    if (context.getConfig('enableFeature')) {
        behaviors.push('new-behavior');
    } else {
        behaviors.push('old-behavior');
    }

   return {
        behaviors: behaviors
   };

});
```

Currently, the Context object method, `getConfig()`, can only read module config after the module instance is created. It is built that way because it caches the configurations after reading (https://github.com/box/t3js/blob/master/lib/application.js#L612). While this is more optimal, it makes configuring dynamic behaviors difficult.

@nzakas - Thoughts?

### PR DESCRIPTION
**Pull request by j3tan**
fixes #143

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @j3tan has signed the CLA. Thanks for the pull request!

File: lib/application.js [hunk 1]
```diff
@@ -604,33 +604,37 @@ Box.Application = (function() {
...
-					return null;
-				} else if (typeof name === 'undefined') {
-					return instanceData.config;
-				} else if (name in instanceData.config) {
-					return instanceData.config[name];
-				} else {
-					return null;
+				// Cache the configurations for performance, if the module instance has been created
+				if (instanceData) {
+					instanceData.config = moduleConfig;
```
Comment (nzakas) [inline lib/application.js]: Do you want to cache even if the instance isn't created yet? (Optimization, not required, just a thought)

**Comment by nzakas** [comment]
+2 just one thought

File: lib/application.js (hunk 1 above)
Comment (j3tan) [inline lib/application.js]: Would require a bit of refactor since instanceData is managed by start/stop. Probably not worth it at this point

**Pull request merged**